### PR TITLE
Gasoline zombies!

### PR DESCRIPTION
**Pull request by ghost**
#### Summary

```SUMMARY: Content "Adds new zombie - Gasoline Zombie"```

#### Purpose of change
Adds a new zombie type - gasoline zombie, in the zombie brute upgrade tree.
It is slow and lumbering, leaks a trail of gasoline, and when it dies it spreads a conflagration of fire

#### Describe the solution
Added a new flag for DRIPS_GASOLINE and mondeath CONFLAGRATION

#### Describe alternatives you've considered
Could be crazy cataclysm only, might be too silly, but I find the idea fun.

#### Additional context
N/A

### CODE REVIEW

File: src/mondeath.cpp [hunk 1]
```diff
@@ -844,3 +844,19 @@ void mdeath::fireball( monster &z )
         normal( z );
     }
 }
+
+
+void mdeath::conflagration( monster &z )
+{
+    for( int x = z.pos().x - 1; x <= z.pos().x + 1; ++x ) {
+        for( int y = z.pos().y - 1; y <= z.pos().y + 1; ++y ) {
```
Comment (codemime) [inline src/mondeath.cpp]: Ther's `map::points_in_radius()` for this purpose.

File: src/mondeath.cpp [hunk 2]
```diff
@@ -844,3 +844,19 @@ void mdeath::fireball( monster &z )
...
+            tripoint fire_tri = tripoint( x, y, z.pos().z );
+            g->m.propagate_field( fire_tri, fd_fire, 18, 3 );
+        }
+    }
+    std::string explode = string_format(
+                              _( "the %s belches sparks and flame erupts from it's stomach!" ),
```
Comment (codemime) [inline src/mondeath.cpp]: I'd rewrite the description: you might not see the eruption itself (only hear it), but the wording implies that you can actually see what's happening. Another option is adding a check like `g->u.sees( z.pos() )`.

File: doc/JSON_FLAGS.md [hunk 3]
```diff
@@ -243,6 +244,7 @@ Flags used to describe monsters and define their properties and abilities.
 - ```DIGS``` Digs through the ground.
 - ```DOGFOOD``` Becomes friendly / tamed with dog food.
 - ```DRIPS_NAPALM``` Occasionally drips napalm on move.
+- ```DROPS_GASLOLINE``` Occasionally drips gasoline on move.
```
Comment (codemime) [inline doc/JSON_FLAGS.md]: Why not `DRIPS`?

File: data/json/monsters.json [hunk 4]
```diff
@@ -411,6 +411,50 @@
...
+      "SMELLS",
+      "STUMBLES",
+      "WARM",
+      "BASHES",
+      "GROUP_BASH",
+      "POISON",
+      "NO_BREATHE",
+      "DRIPS_GASOLINE",
+      "PUSH_MON",
+      "PATH_AVOID_DANGER_1",
```
Comment (codemime) [inline data/json/monsters.json]: This isn't particularly zombish.

File: doc/JSON_FLAGS.md (hunk 3 above)
Comment (davidpwbrown) [reply doc/JSON_FLAGS.md]: oops yeah

File: data/json/monsters.json (hunk 4 above)
Comment (davidpwbrown) [reply data/json/monsters.json]: The flags? they are the same as a boomer?

Comment (codemime) [reply data/json/monsters.json]: I meant specifically `PATH_AVOID_DANGER_1`. I wonder is there a reason why boomers have it?

File: src/mondeath.cpp [hunk 5]
```diff
@@ -844,3 +844,14 @@ void mdeath::fireball( monster &z )
         normal( z );
     }
 }
+
+
+void mdeath::conflagration( monster &z )
+{
+    for( auto &&dest : g->m.points_in_radius( z.pos(), 1 ) ) {
```
Comment (codemime) [inline src/mondeath.cpp]: Why an rvalue reference? I'd replace with a const reference: `for( const auto &dest ..`.

File: src/mondeath.cpp [hunk 6]
```diff
@@ -844,3 +844,14 @@ void mdeath::fireball( monster &z )
...
+        g->m.propagate_field( dest, fd_fire, 18, 3 );
+    }
+    std::string explode = string_format( _( "a %s explode!" ), z.name().c_str() );
```
Comment (codemime) [inline src/mondeath.cpp]: This can also be made const (sorry, missed it the first time).

File: data/json/monsters.json (hunk 4 above)
Comment (davidpwbrown) [reply data/json/monsters.json]: No idea, that's weird, but this isn't really a boomer, so I can remove that to make it more in line with other zombies

**Comment by Mcrone617** [comment]
I feel like this would fit better in the Boomer upgrade tree honestly, as we currently have a armored fire zombie already (the zombie burner). Also like the burner, maybe make it only explode and spread a bunch of fire half the time or something? That way it's kinda risky to kill, but by the time you find one, hopefully you could kite it out onto a road or something and snipe it without burning down the entire town.

File: data/json/monsters.json [hunk 7]
```diff
@@ -411,6 +411,49 @@
       "FILTHY"
     ]
   },
+  {
+    "id": "mon_gas_zombie",
+    "type": "MONSTER",
+    "name": "gasoline zombie",
+    "description": "A huge bloated zombie that appears to have fed upon gasoline, fumes and flames escape from it's mouth and fuel leaks from it's waddling form.",
```
Comment (thquinn) [inline data/json/monsters.json]: "A huge bloated zombie that appears to have fed upon gasoline; fumes and flames escape from its mouth and fuel leaks from its waddling form."

File: data/json/morale_types.json [hunk 8]
```diff
@@ -324,6 +324,11 @@
     "type": "morale_type",
     "text": "Anguished by memories of butchering human corpse"
   },
+  {
+    "id": "morale_pain",
+    "type": "morale_type",
+    "text": "In pain"
```
Comment (codemime) [inline data/json/morale_types.json]: This was committed accidentally, I assume.

Comment (davidpwbrown) [reply data/json/morale_types.json]: dammit, I am on fire tonight with these mistakes. ( hah )

**Comment by davidpwbrown** [comment]
> I feel like this would fit better in the Boomer variant upgrade tree honestly, as we currently have a armored fire zombie already (the zombie burner). Also like the burner, maybe make it only explode and spread a bunch of fire half the time or something? That way it's kinda risky to kill, but by the time you find one, hopefully you could kite it out onto a road or something and snipe it without burning down the entire town.
> 
> Still, its a fun idea, and would passively help reduce loot numbers in later years, as this type of zombie evolves, its more likely that buildings will burn down, which makes it a good idea in that regard.

Yeah I can do that tomorrow,  the zombie burner, I've never ever actually seen burn, one in ten seems too low, but a guaranteed fireball is perhaps unfair, yes. 
So 50% of the time seems fair. 

Boomers don't currently have an upgrade group, but I could add that.

File: data/json/monsters.json [hunk 9]
```diff
@@ -411,6 +411,49 @@
       "FILTHY"
     ]
   },
+  {
+    "id": "mon_gas_zombie",
```
Comment (Night-Pryanik) [inline data/json/monsters.json]: I suggest renaming it to `mon_zombie_gas`, as other zombies are named `mon_zombie_xxx`.

File: src/mondeath.cpp [hunk 10]
```diff
@@ -844,3 +844,14 @@ void mdeath::fireball( monster &z )
...
     }
 }
+
+
+void mdeath::conflagration( monster &z )
+{
+    for( const auto &dest : g->m.points_in_radius( z.pos(), 1 ) ) {
+        g->m.propagate_field( dest, fd_fire, 18, 3 );
+    }
+    const std::string explode = string_format( _( "a %s explode!" ), z.name().c_str() );
```
Comment (Night-Pryanik) [inline src/mondeath.cpp]:
```suggestion
    const std::string explode = string_format( _( "a %s explode!" ), z.name() );
```

File: src/mondeath.h [hunk 11]
```diff
@@ -41,6 +41,7 @@ void detonate( monster &z );      // Take the enemy with you
 void broken_ammo( monster &z );   // Breaks ammo and then itself
 void preg_roach( monster &z );    // Spawns 1-3 roach nymphs
 void fireball( monster &z );      // Explodes in fire
+void conflagration( monster &z ); // Similar to above but bigger and gauranteed.
```
Comment (Night-Pryanik) [inline src/mondeath.h]:
```suggestion
void conflagration( monster &z ); // Similar to above but bigger and guaranteed.
```

File: doc/JSON_FLAGS.md [hunk 12]
```diff
@@ -243,6 +244,7 @@ Flags used to describe monsters and define their properties and abilities.
 - ```DIGS``` Digs through the ground.
 - ```DOGFOOD``` Becomes friendly / tamed with dog food.
 - ```DRIPS_NAPALM``` Occasionally drips napalm on move.
+- ```DRIPS_GASLOLINE``` Occasionally drips gasoline on move.
```
Comment (matskuman5) [inline doc/JSON_FLAGS.md]:
```suggestion
- ```DRIPS_GASOLINE``` Occasionally drips gasoline on move.
```

Comment (davidpwbrown) [reply doc/JSON_FLAGS.md]: I Really do lack quality control dont I.
however, gasloline is a good typo.

**Comment by DracoGriffin** [comment]
There are tons of suggestions and feedback that should be considered from #23729. More specifically, how does this address [concerns I had](https://github.com/CleverRaven/Cataclysm-DDA/pull/23729#issuecomment-389612909) for the other "gasoline zombie"?

And I'm not sure we'll have the benefit of Leland doing extensive testing as he had in #23729.

Lastly, it seems to just re-use mechanisms @Night-Pryanik established initially with duplicate code?

I'm normally one to encourage all types of contributions and additions, but given the effects of fire (and fungal, but that's offtopic), it can become extremely severe for player save games and cause slowdowns. Having monsters like these that populate cities and labs and other areas so prevalently, it can could easily lead to issues with players swarming Issues about unknown slowdowns or bug reports/complaints on Discord/Discourse/Reddit. Especially with Z-Levels enabled, players starting in labs and experiencing heavy slowdowns because a gasoline zombie(s) started major fires on the surface or vice versa.

If anything, I'd rather see the gasoline zombie, when killed, explode and cover a small area with gasoline (similar to vehicles with damaged gas tanks) but if player is in range, all their clothes become drenched in gasoline --- causing some type of weakness (the smell causing lightheadedness and such), and when near fires, will catch on fire. The only way to rid of the weakness would be to remove the clothes (making "WATERPROOF" flag more valuable as it would prevent inner clothing from being affected) until cleaned by washing machine and/or washboard+soap. The gasoline could cause a similar flag like "(filthy)" but with "(gasoline)" or something ("covered with gasoline" may be a bit too long).

**Comment by Ilysen** [comment]
Among other things, I don't see any real reason why gas would ignite just by existing, and the zombie itself isn't fireproof, so having it visibly exhaling flames seems silly to me. It also doesn't make sense, to me, to explode on death, for the same reasons that the burner does. If you beat something to death with a baseball bat, there's *nothing* causing the flammable fuel to spark. I could understand if it was through bullet deaths only, but it's not.

The zombie burner was tested to have almost this exact same function, and caused mass destruction. It was only remedied because they only spawn in military pools, so they don't populate towns and the like.

Fire is very, *very* destructive, and I think in their current version they would destroy entire towns and cities by existing. In its current state, I have to disagree with having it added.

**Comment by cosmo-naut** [comment]
Could be useful as a late game way to get more gasoline, as currently only diesel is 'renewable'

**Comment by Firestorm01X2** [comment]
It is better to merge this first:
https://github.com/CleverRaven/Cataclysm-DDA/pull/28797

It fixes:
https://github.com/CleverRaven/Cataclysm-DDA/issues/23063

People are starting editing spawn lists freqently with small chunks. My PR edits a lot of them. 

My PR is ready.

**Comment by Ilysen** [comment]
Would it be too kneejerk to put this into the explosive zombies blacklist?

**Comment by DracoGriffin** [comment]
> Would it be too kneejerk to put this into the explosive zombies blacklist?

No. Please feel free to update the blacklist mod.

**Comment by davidpwbrown** [comment]
to be honest I was undecided as to just close this silly throwaway idea I had or put it into crazy cataclysm, and at least enact the changes suggested by review, but didn't get to it quick enough I guess! ...

**Pull request merged**